### PR TITLE
Fixes #19248: Empty /var/rudder/policy-generation-info/node-configuration-hashes.json after a policy generation that changed nothing

### PR DESCRIPTION
**Pull request by fanf**
https://issues.rudder.io/issues/19248

### CODE REVIEW

File: webapp/sources/rudder/rudder-core/src/test/scala/com/normation/rudder/services/policies/nodeconfig/NodeConfigurationCacheRepositoryTest.scala [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+
+  val h3_0 = NodeConfigurationHash(NodeId("node3"), d1, 0, 0, 0, Set(
+    PolicyHash(PolicyId(RuleId("r3"), DirectiveId("d3"), TechniqueVersion("1.0")), 0)
+  ))
+
+  val s1 = Set(h1_0, h0_0, h2_0)
+
+  sequential
+
+  "when nothing was written, we should get an empty succe" >> {
```
Comment (ncharles) [inline webapp/sources/rudder/rudder-core/src/test/scala/com/normation/rudder/services/policies/nodeconfig/NodeConfigurationCacheRepositoryTest.scala:108]:
```suggestion
  "when nothing was written, we should get an empty success" >> {
```

File: webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/nodeconfig/NodeConfigurationCacheRepository.scala [hunk 2]
```diff
@@ -543,11 +543,17 @@ class FileBasedNodeConfigurationHashRepository(path: String) extends NodeConfigu
...
-        _       <- nonAtomicWrite(NodeConfigurationHashes(hashes.toList.sortBy(_.id.value)))
-      } yield hashes.map(_.id)
-    }).toBox
+    val nodeIds = hashes.map(_.id)
+    timeLog(s => s"Updating node configuration hashes took ${s}")(
+      if(hashes.size == 0) nodeIds.succeed
+      else semaphore.withPermits(1) {
+        for {
+          current  <- nonAtomicRead()
+          filtered =  current.hashes.filterNot(h => nodeIds.contains(h.id))
```
Comment (ncharles) [inline webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/nodeconfig/NodeConfigurationCacheRepository.scala:552]: i'm unsure tht you need this one: you are traversing current.hashes , and n times the nodeIds

doing simply (filtered ++ hashes).toList.sortBy(_.id.value) would yield the same result

**Review by ncharles — COMMENTED** [review]
i *think* the filterNot is overcomplicating the method

**Comment by fanf** [comment]
PR updated with a new commit

**Review by ncharles — APPROVED** [review]

File: webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/nodeconfig/NodeConfigurationCacheRepository.scala (hunk 2 above)
Comment (fanf) [reply webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/nodeconfig/NodeConfigurationCacheRepository.scala:552]: no, `filtered` is a list, not a map, and `hashes` is not a map either, so I will have several node config hashes for the same node Id.  
I would need to do `current.hashes.map(h => (h.id, h)).toMap`, same for `hashes`, so two traverse, and one more (more coslty) for the map merge.

**Comment by fanf** [comment]
OK, squash merging this PR

**Pull request merged**